File: demos/demo_multiple_scenes.cpp
/*******************************************************************************
 * Demo: Multiple Scenes
 *
 * 2×3 grid of scenes, each showing a triangle in a different color.
 *
 * Demonstrates: App::setSceneGrid, App::sceneCamera, multiple scenes.
 *
 * Controls: H for help
 *******************************************************************************/

#include "geodraw/app/app.hpp"
#include "geodraw/geometry/geometry.hpp"

using namespace geodraw;

static const glm::vec3 COLORS[6] = {
    Colors::RED,    Colors::GREEN,   Colors::BLUE,
    Colors::ORANGE, Colors::CYAN,    Colors::MAGENTA,
};
static const Triangle3 TRIANGLE{Pos3{1, 1, 0}, Pos3{3, 2, 0}, Pos3{1, 2, 1}};
static const Pose3 AXES;

int main() {
    App app(1280, 720, "Demo Multiple Scenes");

    for (int i = 0; i < 6; ++i) app.addScene();
    app.setSceneGrid(2, 3);

    // Set a static camera for each sub-scene
    for (int i = 0; i < 6; ++i) {
        app.sceneCamera(i).autoFrame(
            glm::vec3(1.0f, 1.0f, 0.0f),
            glm::vec3(3.0f, 2.0f, 1.0f));
    }

    app.addUpdateCallback([&](float) {
        for (int i = 0; i < 6; ++i) {
            app.scene(i).clear();
            app.scene(i).Add(TRIANGLE, 2.0f, LineStyle::Solid, 1.0f, COLORS[i]);
            app.scene(i).Add(AXES, 2.0f, 5.0f);
        }
    });

    app.run();
    return 0;
}

File: python/demos/demo_multiple_scenes.py
"""Demo: Multiple Scenes (Python)

2×3 grid of scenes, each showing a triangle in a different color.
Demonstrates: App.set_scene_grid, App.auto_frame_scene, multiple scenes.
"""
import geodraw

COLORS = [
    (1.0, 0.2, 0.2), (0.2, 1.0, 0.2), (0.2, 0.2, 1.0),
    (1.0, 0.6, 0.1), (0.1, 0.8, 0.8), (0.8, 0.1, 0.8),
]
TRIANGLE = [(1, 1, 0), (3, 2, 0), (1, 2, 1)]


def main():
    app = geodraw.App(1280, 720, "Demo Multiple Scenes")

    for _ in range(6):
        app.add_scene()
    app.set_scene_grid(2, 3)

    for i in range(6):
        app.auto_frame_scene(i, (1, 1, 0), (3, 2, 1))

    @app.add_update_callback()
    def update(dt):
        for i in range(6):
            s = app.scene(i)
            s.clear()
            s.add_triangle(TRIANGLE, color=COLORS[i])
            s.add_pose((0, 0, 0), [[1, 0, 0], [0, 1, 0], [0, 0, 1]], 2.0, 5.0);

    app.run()


if __name__ == "__main__":
    main()
